### DISCUSSION TITLE
Publish unist packages and friends

### DISCUSSION THREAD

**Discussion by remcohaszing** [syntax-tree/unist#32] [Ideas] [answered]
### Problem

Type definitions for various *st packages are published using DefinitelyTyped. Having these types packages without an accompanying source package on DefinitelyTypes has caused various issues for me when working with unist ASTs. Just some examples:

The following are some issues related to resolving packages:

- `eslint-plugin-import` doesn’t **just work**. It requires the need to use [`eslint-import-resolver-typescript`](https://www.npmjs.com/package/eslint-import-resolver-typescript).
- Even if `eslint-import-resolver-typescript` is configured, some rules from `eslint-plugin-node` don’t just work either.

Also TypeScript itself has issues with this. For example, the following code:

```ts
import { Node } from 'unist';

class Foo {
  @Bar
  node: Node;
}
```

Gets transpiled to:

```js
var __decorate = /* truncated */;
var __metadata = /* truncated */;
var _a;
import { Node } from 'unist';
class Foo {
}
__decorate([
    Bar,
    __metadata("design:type", typeof (_a = typeof Node !== "undefined" && Node) === "function" ? _a : Object)
], Foo.prototype, "node", void 0);
```

So `Node` will be imported from `unist`, even though the `unist` package doesn’t exist.

This can be worked around by using a type import, but the generated error will be unclear.

These are just the sutiations I recall from the top of my head. There are more situations.

### Solution

I propose to publish all syntax tree AST type definition packages using their own name, pulling them from DefinitelyTypes. This will resolve some of the issues.

To resolve more issues, I propose to add an empty `index.cjs` and `index.mjs` file to each package. I.e. `import { Node } from 'unist'` will just result in `Node` being undefined, not in an import error.

### Versioning

Some packages haven’t even been published to `npm` (https://www.npmjs.com/package/xast). For these packages it the version doesn’t really matter.

**Side note**: This also means anyone could just publish a package named `xast`.

Other packages have been published before (https://www.npmjs.com/package/mdast). The new version should have the major version bumped compared to the last release.

**Comment by ChristianMurphy** [comment]
Speaking to this specific issue

```typescript
import { Node } from 'unist';
```

transpiled to try to import a value

https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-8.html#type-only-imports-and-export resolves this

**Reply by remcohaszing** [reply]
True, but still I’d prefer `Node` to be undefined over having an import error.

**Reply by wooorm** [reply]
It sounds like TS has a suggested way to solve type-only things. Why should we deviate from that?

**Comment by wooorm** [comment]
I don’t think TS-only unist/mdast/hast/xast/nlcst/esast packages makes sense: they’re a contract on an interface.
But types are accepted in all other unified packages though

**Reply by remcohaszing** [reply]
This isn’t unique to unist. More types-only packages exist.

- https://www.npmjs.com/package/openapi-types
- https://www.npmjs.com/package/type-fest
- https://www.npmjs.com/package/@fortawesome/fontawesome-common-types
- https://www.npmjs.com/package/@sentry/types

**Reply by ChristianMurphy** [reply]
this was somewhat explored in https://github.com/syntax-tree/unist-ts
but never came to fruition https://github.com/syntax-tree/unist-ts/issues/4

**Comment by ChristianMurphy** [comment, marked as answer]
> To resolve more issues, I propose to add an empty index.cjs and index.mjs file to each package. I.e. import { Node } from 'unist' will just result in Node being undefined, not in an import error.

I think @wooorm has `mdast` https://www.npmjs.com/package/mdast and `hast` https://www.npmjs.com/package/hast published to npm with an index.js file, since the content doesn't seem to matter in this case, would that be enough?
(`unist`, `xast`, and `dotast` are still unpublished and an open question)

**Reply by wooorm** [reply]
I have mdast and hast because those used to be where remark/rehype lived, before the re* rebranding and separate syntax tree documents.

esast would be a problem.

**Reply by remcohaszing** [reply]
This shows that even if the types exist on DefinitelyTyped, the package name should be reserved by unified to prevent name clashes. `esast` appears to be an abandoned project, but it might as well have been a project that’s used and people want to consume in TypeScript.

I believe `npm` has a policy where such package names can be claimed. (Although asking the current owner is nicer of course :smile: )

**Reply by wooorm** [reply]
I interpreted the other way around: that we can pick names for syntax trees even when they’re squatted on npm.

But indeed, we could ask the owner and/or npm. I have some experience there!

**Reply by remcohaszing** [reply]
An alternative could be to use scoped packages. Then either each syntax tree could be registered as a scope, or the `syntax-tree` scope could be used.

I.e.

- `@syntax-tree/esast`
- `@syntax-tree/hast`
- `@syntax-tree/mdast`
- `@syntax-tree/xast`
- etc…

Or

- `@esast/types`
- `@hast/types`
- `@mdast/types`
- `@xast/types`
- etc…

If the `syntax-tree` scope would be used, there’s no chance of name clashes.

If a scope is used per category, these scopes can be used to register more official packages. I.e. `@remark/lint`, `@remark/parse`, etc. I like this, but I recall reading at some point you didn’t like this when someone else proposed it.

**Reply by wooorm** [reply]
using scopes creates a difference between `@remark/thing` and `remark-thing`: one is seen as official, but I think it’s good that all plugins share that level of “officialness”.
For these interfaces, I’m fine with it.
And we do have npm organizations with the same names as GH orgs, which gives us those names as scopes, so we do have `@syntax-tree/...`

**Reply by wooorm** [reply]
@ChristianMurphy What do you think about shipping them as such?

**Reply by wooorm** [reply]
another alternative is `unist-types` ,`mdast-types`, etc...

**Reply by ChristianMurphy** [reply]
:thinking: many packages' typings depend on `@types/unist`, `@types/mdast`, and `@types/hast` already.
If we rename them, whether to a scoped package or to `*-types` we'd need to refactor typings in a lot of packages.

Some options include:
1. publishing directly to `unist`, `mdast`, and `hast`
2. re-exporting `unist`, `mdast`, and `hast` under another name like `unist-types`, etc
3. move the packages and refactor all the unified typings
4. create an ESLint preset which ignores these module names
5. update the eslint plugin upstream to handle a package with a definitely typed package, but no unscoped package

Personally I tend to lean toward 4 or 5 first and foremost, the issues described feel like they are mostly toolchain related, and that resolving it outside the toolchain is a workaround.
With 1 or 2 being the next preference, as they wouldn't require refactoring typings.

**Reply by wooorm** [reply]
1: The drawback here is that I feel `unist` etc are specs/documentation. That could also work in other type or programming languages. So typescript types are one implementation, but there could be more. So I feel that it’s a bit weird to tie their names to typescript types.
2: You mean that the types would still be maintained in DT, but exported for convenience with a different name?
3. This has my preference
4/5: What would the ESLint preset/plugin do/solve?

For the name, I guess the original idea of using `unist-ts` etc is fine as well.

Still, DT is a pretty good place to handle this all, and it sounds like the original issue [has a solution](https://github.com/syntax-tree/unist/discussions/32#discussioncomment-379174)

**Reply by ChristianMurphy** [reply]
> 2. You mean that the types would still be maintained in DT, but exported for convenience with a different name?

Exactly, the package would just be `export * from '<definitely-typed-package>'`

> 4/5: What would the ESLint preset/plugin do/solve?

It would stop errors from being thrown for importing types only packages from definitely typed.

**Reply by wooorm** [reply]
But 4/5 would not solve the TS compile error also mentioned in the original post, correct? Or can that be solved in another way?

**Reply by remcohaszing** [reply]
> 1. publishing directly to `unist`, `mdast`, and `hast`

This has my preference. This causes the least amount of issues for working with these types.

> 1. The drawback here is that I feel `unist` etc are specs/documentation. That could also work in other type or programming languages. So typescript types are one implementation, but there could be more. So I feel that it’s a bit weird to tie their names to typescript types.

This assumes the types **packages** would belong in the existing **repositories**. A package is not the same as a repository. I.e. a package named `unist` can exist in a repository named `https://github.com/syntax-tree/unist-types`

> 2. re-exporting unist, mdast, and hast under another name like unist-types, etc

I think this indirection confusing at most, and it just adds another dependency in the user’s `node_module` directory for no real benefit.

> 3. move the packages and refactor all the unified typings

This has my preference, however, considering the required refactors I don’t know if this is worth the trouble.

> 4. create an ESLint preset which ignores these module names

I don’t think this can be solved this way.

> 5. update the eslint plugin upstream to handle a package with a definitely typed package, but no unscoped package

Perhaps this is the right thing to do. I.e. even if `mdast` is moved and some errors are gone, still https://github.com/remcohaszing/remark-mermaidjs/blob/master/src/index.ts#L5 would cause `eslint-plugin-import/no-extraneous-dependencies` to complain, even though this import is nowhere to be found in the transpiled JavaScript or types definitions.

**Reply by ChristianMurphy** [reply]
> The drawback here is that I feel unist etc are specs/documentation

I agree, but I also tend to see typescript as specs/documentation. :slightly_smiling_face: 

> 1 publishing directly to unist, mdast, and hast
> 5 update the eslint plugin upstream to handle a package with a definitely typed package, but no unscoped package

these two are the ones I lean toward

> 3 move the packages and refactor all the unified typings

I'd really like to avoid this, there are a lot of packages with typings within unified.
and even more within the community that would need to migrate.
It might be okay as part of a larger effort like https://github.com/unifiedjs/rfcs/pull/5, but by itself, it seems like a lot of work for not a lot of benefit.

> > 4 create an ESLint preset which ignores these module names
>
> I don’t think this can be solved this way.

I think it _can_, https://github.com/benmosher/eslint-plugin-import/blob/HEAD/docs/rules/no-unresolved.md#ignore
That doesn't mean it _should_ be solved this way, but it is possible.

**Reply by remcohaszing** [reply]
> I think it can, https://github.com/benmosher/eslint-plugin-import/blob/HEAD/docs/rules/no-unresolved.md#ignore

I totally missed this somehow!

I don’t think it makes sense to create an ESLint configuration just for this purpose. but I’ll add it to my own shareable config then. I think this solves the biggest issue I’m having currently.

**Reply by wooorm** [reply]
Oh nice!

Well, uhmm, does that mean we can close this issue, because DT is fine if ESLint is configured?

**Reply by remcohaszing** [reply]
Yup :+1:

**Comment by wooorm** [comment]
Could this be solved in DT? Can `index.js` files be added there?

**Reply by remcohaszing** [reply]
No. If the TypeScript decorator causes the import to stay, then the runtime (NodeJS) will try to import `mdast`, which doesn’t resolve to `@types/mdast`.
